src/main/java/teammates/logic/core/CoursesLogic.java

```diff
@@ -719,4 +719,31 @@ public boolean isCourseArchived(CourseAttributes course, InstructorAttributes in
         boolean isCourseArchived = (instructor.isArchived != null) ? instructor.isArchived : course.isArchived;
         return isCourseArchived;
     }
-}
+    
+    // TODO: Optimize extractActiveCourses() and extractArchivedCourses() to reduce the number of repeated calls of
+    // isCourseArchived(), which retrieves information from the database
+    
+    public List<CourseDetailsBundle> extractActiveCourses(List<CourseDetailsBundle> courseBundles, String googleId) {
+        CourseDetailsBundle.sortDetailedCoursesByCourseId(courseBundles);
+        List<CourseDetailsBundle> result = new ArrayList<CourseDetailsBundle>();
+        
+        for (CourseDetailsBundle courseBundle : courseBundles) {
+            if (!isCourseArchived(courseBundle.course.id, googleId)) {
+                result.add(courseBundle);
+            }
+        }
+        return result;
+    }
+    
+    public List<CourseDetailsBundle> extractArchivedCourses(List<CourseDetailsBundle> courseBundles, String googleId) {
+        CourseDetailsBundle.sortDetailedCoursesByCourseId(courseBundles);
+        List<CourseDetailsBundle> result = new ArrayList<CourseDetailsBundle>();
```

Comment:
Should these methods be in `CourseLogic` instead of `*PageAction`?
Are there multiple classes requesting them separated by whether they are active or archived?